tests/test_charm.py

```diff
@@ -51,6 +51,48 @@ def test_website_relation_joined(self, open, ingress_address):
         self.assertEqual(data["private-address"], "192.168.1.17")
         self.assertEqual(data["port"], '17070')
 
+    @patch("builtins.open", new_callable=mock_open, read_data=agent_conf)
+    def test_metrics_endpoint_relation(self, open):
+        harness = Harness(JujuControllerCharm)
+        self.addCleanup(harness.cleanup)
+        harness.begin()
+
+        harness.add_network(address="192.168.1.17", endpoint="metrics-endpoint")
+
+        mock_opener = mockOpener(self)
+        harness.charm.control_socket.opener = mock_opener
+
+        mock_opener.expect(
+            url='http://localhost/metrics-users',
+            method='POST',
+            body_regex=r'{"username": "juju-metrics-r\d+", "password": ".*"}',
+            response=mockResponse(
+                headers=mockHeaders(content_type='application/json'),
+                body='{"message":"created user \\"juju-metrics-r0\\""}'
+            )
+        )
+        relation_id = harness.add_relation('metrics-endpoint', 'prometheus-k8s')
+
+        # Check metrics endpoint provider config
+        scrape_config = harness.charm.metrics_endpoint._scrape_jobs[0]
+        self.assertEqual(scrape_config['metrics_path'], '/introspection/metrics')
+        self.assertEqual(scrape_config['static_configs'], [{'targets': ['*:17070']}])
+        self.assertEqual(scrape_config['scheme'], 'https')
+        self.assertEqual(scrape_config['basic_auth']['username'], 'user-juju-metrics-r0')
+        self.assertEqual(scrape_config['tls_config']['ca_file'], 'fake')
+        self.assertEqual(scrape_config['tls_config']['server_name'], 'juju-apiserver')
+
+        # Test removing the relation
+        mock_opener.expect(
+            url=f'http://localhost/metrics-users/juju-metrics-r{relation_id}',
+            method='DELETE',
+            response=mockResponse(
+                headers=mockHeaders(content_type='application/json'),
+                body='{"message":"deleted user \\"juju-metrics-r0\\""}'
+            )
+        )
+        harness.remove_relation(relation_id)
```

Comment:
After this, we should test that the request actually went through. Per above, if we change to mocking at the `remove_metrics_user` level, this is simply a matter of `mock.assert_called_once_with(expected_username)`.